File: lsif/src/shared/store/uploads.ts
import * as pgModels from '../models/pg'
import { Brackets, Connection, EntityManager } from 'typeorm'
import { FORMAT_TEXT_MAP, Span, Tracer } from 'opentracing'
import { instrumentQuery, withInstrumentedTransaction } from '../database/postgres'
import { PlainObjectToDatabaseEntityTransformer } from 'typeorm/query-builder/transformer/PlainObjectToDatabaseEntityTransformer'
import { Logger } from 'winston'

export interface LsifUploadWithPlaceInQueue extends pgModels.LsifUpload {
    placeInQueue: number | null
}

/**
 * A wrapper around the database tables that control uploads. This class has
 * behaviors to enqueue uploads and dequeue them for the processor to convert
 * in a transactional manner.
 */
export class UploadManager {
    /**
     * Create a new `UploadManager` backed by the given database connection.
     *
     * @param connection The Postgres connection.
     */
    constructor(private connection: Connection) {}

    /**
     * Get the count of uploads in the given state.
     *
     * @param state The state.
     * @param entityManager An entity manager to use if within a transaction.
     */
    public getCount(
        state: string,
        entityManager: EntityManager = this.connection.createEntityManager()
    ): Promise<number> {
        return entityManager
            .getRepository(pgModels.LsifUpload)
            .createQueryBuilder()
            .select()
            .where({ state })
            .getCount()
    }

    /**
     * Get the uploads in the given state.
     *
     * @param repositoryId The repository identifier.
     * @param state The state.
     * @param query A search query.
     * @param visibleAtTip If true, only return dumps visible at tip.
     * @param limit The maximum number of uploads to return.
     * @param offset The number of uploads to skip.
     */
    public async getUploads(
        repositoryId: number,
        state: pgModels.LsifUploadState | undefined,
        query: string,
        visibleAtTip: boolean,
        limit: number,
        offset: number
    ): Promise<{ uploads: LsifUploadWithPlaceInQueue[]; totalCount: number }> {
        const { uploads, raw, totalCount } = await instrumentQuery<{
            uploads: pgModels.LsifUpload[]
            raw: { upload_id: number; rank: string | undefined }[]
            totalCount: number
        }>(async () => {
            let queryBuilder = this.connection
                .getRepository(pgModels.LsifUpload)
                .createQueryBuilder('upload')
                .addSelect('ranked.rank', 'rank')
                .leftJoin(
                    qb =>
                        qb
                            .subQuery()
                            .select('ranked.id, RANK() OVER (ORDER BY ranked.uploaded_at) as rank')
                            .from(pgModels.LsifUpload, 'ranked')
                            .where("ranked.state = 'queued'"),
                    'ranked',
                    'ranked.id = upload.id'
                )
                .where({ repositoryId })
                .orderBy('uploaded_at', 'DESC')
                .limit(limit)
                .offset(offset)

            if (state) {
                queryBuilder = queryBuilder.andWhere('state = :state', { state })
            }

            if (query) {
                const clauses = ['commit', 'root', 'indexerName', 'failure_summary', 'failure_stacktrace'].map(
                    field => `"${field}" LIKE '%' || :query || '%'`
                )

                queryBuilder = queryBuilder.andWhere(
                    new Brackets(qb =>
                        clauses.slice(1).reduce((ob, c) => ob.orWhere(c, { query }), qb.where(clauses[0], { query }))
                    )
                )
            }

            if (visibleAtTip) {
                queryBuilder = queryBuilder.andWhere('visible_at_tip = true')
            }

            const [{ entities, raw: rawEntities }, count] = await Promise.all([
                queryBuilder.getRawAndEntities(),
                queryBuilder.getCount(),
            ])

            return { uploads: entities, raw: rawEntities, totalCount: count }
        })

        const ranks = new Map(raw.map(r => [r.upload_id, parseInt(r.rank || '', 10)]))
        return { uploads: uploads.map(u => ({ ...u, placeInQueue: ranks.get(u.id) || null })), totalCount }
    }

    /**
     * Get an upload by identifier.
     *
     * @param id The upload identifier.
     */
    public getUpload(id: number): Promise<LsifUploadWithPlaceInQueue | undefined> {
        return withInstrumentedTransaction(this.connection, async () => {
            const {
                entities,
                raw,
            }: { entities: pgModels.LsifUpload[]; raw: { rank: string | null }[] } = await this.connection
                .getRepository(pgModels.LsifUpload)
                .createQueryBuilder('upload')
                .addSelect('ranked.rank', 'rank')
                .leftJoin(
                    qb =>
                        qb
                            .subQuery()
                            .select('ranked.id, RANK() OVER (ORDER BY ranked.uploaded_at) as rank')
                            .from(pgModels.LsifUpload, 'ranked')
                            .where("ranked.state = 'queued'"),
                    'ranked',
                    'ranked.id = upload.id'
                )
                .where({ id })
                .limit(1)
                .getRawAndEntities()

            if (entities.length === 0) {
                return undefined
            }

            return { ...entities[0], placeInQueue: parseInt(raw[0].rank || '', 10) || null }
        })
    }

    /**
     * Delete an upload. This returns true if the upload existed. Also remove referenced
     * package and reference rows if the upload was successfully processed.
     *
     * Does not delete the file on disk directly. This will be cleaned up later as part
     * of the `removeDeadDumps` function performed at the start of the `purgeOldDumps`
     * task that is run on a schedule in the server context.
     *
     * @param id The upload identifier.
     * @param updateVisibility A function that updates the dumps visible at the tip for
     *     the given repository. This is called if the deleted dump was visible at tip,
     *     as a previously non-visible dump may become visible after deletion.
     */
    public async deleteUpload(
        id: number,
        updateVisibility: (entityManager: EntityManager, repositoryId: number) => Promise<void>
    ): Promise<boolean> {
        return withInstrumentedTransaction(this.connection, async entityManager => {
            const [affected, numAffected]: [
                { repository_id: number; visible_at_tip: boolean }[],
                number
            ] = await instrumentQuery(() =>
                entityManager.query('DELETE FROM lsif_uploads WHERE id = $1 RETURNING repository_id, visible_at_tip', [
                    id,
                ])
            )

            if (numAffected === 0) {
                return false
            }

            if (affected[0].visible_at_tip) {
                await updateVisibility(entityManager, affected[0].repository_id)
            }

            return true
        })
    }

    /**
     * Remove all uploads that are older than `maxAge` seconds. Returns the count of deleted uploads.
     *
     * @param maxAge The maximum age for an upload.
     */
    public async clean(maxAge: number): Promise<number> {
        return (
            (
                await instrumentQuery(() =>
                    this.connection
                        .getRepository(pgModels.LsifUpload)
                        .createQueryBuilder()
                        .delete()
                        .where("state != 'completed'")
                        .andWhere("uploaded_at < now() - (:maxAge * interval '1 second')", { maxAge })
                        .execute()
                )
            ).affected || 0
        )
    }

    /**
     * Move all processing uploads started more than `maxAge` seconds ago that are not currently
     * locked back to the `queued` state.
     *
     * @param maxAge The maximum age for an unlocked upload in the `processing` state.
     */
    public async resetStalled(maxAge: number): Promise<number[]> {
        const results: [{ id: number }[]] = await instrumentQuery(() =>
            this.connection.query(
                `
                    UPDATE lsif_uploads u SET state = 'queued', started_at = null WHERE id = ANY(
                        SELECT id FROM lsif_uploads
                        WHERE state = 'processing' AND started_at < now() - ($1 * interval '1 second')
                        FOR UPDATE SKIP LOCKED
                    )
                    RETURNING u.id
                `,
                [maxAge]
            )
        )

        return results[0].map(r => r.id)
    }

    /**
     * Create a new uploaded with a state of `queued`.
     *
     * @param args The upload payload.
     * @param tracer The tracer instance.
     * @param span The parent span.
     */
    public async enqueue(
        {
            repositoryId,
            commit,
            root,
            filename,
            indexer,
        }: {
            /** The repository identifier. */
            repositoryId: number
            /** The commit. */
            commit: string
            /** The root. */
            root: string
            /** The filename. */
            filename: string
            /** The indexer binary name that produced this dump as specified by the metadata. */
            indexer: string
        },
        tracer?: Tracer,
        span?: Span
    ): Promise<pgModels.LsifUpload> {
        const tracing = {}
        if (tracer && span) {
            tracer.inject(span, FORMAT_TEXT_MAP, tracing)
        }

        const upload = new pgModels.LsifUpload()
        upload.repositoryId = repositoryId
        upload.commit = commit
        upload.root = root
        upload.indexer = indexer
        upload.filename = filename
        upload.tracingContext = JSON.stringify(tracing)
        await instrumentQuery(() => this.connection.createEntityManager().save(upload))

        return upload
    }

    /**
     * Lock and convert a queued upload. If the conversion function throws an error, then
     * the error summary and stack trace will be written to the upload record and the state
     * will be set to "errored".
     *
     * The convert callback is invoked with the locked upload record and the entity manager
     * that locked the record. The callback should use it to operate in the same transaction.
     *
     * This method does NOT mark the upload as complete and the callback MUST be sure to call
     * the `markComplete` method on successful processing. Otherwise the record will block the
     * head of the queue by being re-processed ad nauseam.
     *
     * @param convert The function to call with the locked upload.
     * @param logger The logger instance.
     */
    public async dequeueAndConvert(
        convert: (upload: pgModels.LsifUpload, entityManager: EntityManager) => Promise<void>,
        logger: Logger
    ): Promise<boolean> {
        // First, we select the next oldest upload with a state of `queued` and set
        // its state to `processing`. We do this outside of a transaction so that this
        // state transition is visible to the API. We skip any locked rows as they are
        // being handled by another dump processor.
        const lockResult: [{ id: number }[]] = await this.connection.query(`
            UPDATE lsif_uploads u SET state = 'processing', started_at = now() WHERE id = (
                SELECT id FROM lsif_uploads
                WHERE state = 'queued'
                ORDER BY uploaded_at
                FOR UPDATE SKIP LOCKED LIMIT 1
            )
            RETURNING u.id
        `)
        if (lockResult[0].length === 0) {
            return false
        }
        const uploadId = lockResult[0][0].id

        return withInstrumentedTransaction(this.connection, async entityManager => {
            const results: object[] = await entityManager.query(
                'SELECT * FROM lsif_uploads WHERE id = $1 FOR UPDATE LIMIT 1',
                [uploadId]
            )
            if (results.length === 0) {
                // Record was deleted in race, retry
                return this.dequeueAndConvert(convert, logger)
            }

            // Transform locked result into upload entity
            const repo = entityManager.getRepository(pgModels.LsifUpload)
            const meta = repo.manager.connection.getMetadata(pgModels.LsifUpload)
            const transformer = new PlainObjectToDatabaseEntityTransformer(repo.manager)
            const upload = (await transformer.transform(results[0], meta)) as pgModels.LsifUpload

            try {
                await convert(upload, entityManager)
            } catch (error) {
                logger.error('Failed to convert upload', { error })

                await entityManager.query(
                    `
                        UPDATE lsif_uploads
                        SET state = 'errored', finished_at = now(), failure_summary = $2, failure_stacktrace = $3
                        WHERE id = $1
                    `,
                    [uploadId, error?.message, error?.stack]
                )
            }

            return true
        })
    }

    /**
     * Mark an upload as complete and set its finished timestamp.
     *
     * @param upload The upload.
     * @param entityManager The EntityManager to use as part of a transaction.
     */
    public markComplete(
        upload: pgModels.LsifUpload,
        entityManager: EntityManager = this.connection.createEntityManager()
    ): Promise<void> {
        return entityManager.query("UPDATE lsif_uploads SET state = 'completed', finished_at = now() WHERE id = $1", [
            upload.id,
        ])
    }
}

File: cmd/precise-code-intel-api-server/internal/db/uploads.go
package db

import (
	"context"
	"time"

	"github.com/keegancsmith/sqlf"
)

// StalledUploadMaxAge is the maximum allowable duration between updating the state of an
// upload as "processing" and locking the upload row during processing. An unlocked row that
// is marked as processing likely indicates that the worker that dequeued the upload has died.
// There should be a nearly-zero delay between these states during normal operation.
const StalledUploadMaxAge = time.Second * 5

// Upload is a subset of the lsif_uploads table and stores both processed and unprocessed
// records.
type Upload struct {
	ID                int        `json:"id"`
	Commit            string     `json:"commit"`
	Root              string     `json:"root"`
	VisibleAtTip      bool       `json:"visibleAtTip"`
	UploadedAt        time.Time  `json:"uploadedAt"`
	State             string     `json:"state"`
	FailureSummary    *string    `json:"failureSummary"`
	FailureStacktrace *string    `json:"failureStacktrace"`
	StartedAt         *time.Time `json:"startedAt"`
	FinishedAt        *time.Time `json:"finishedAt"`
	TracingContext    string     `json:"tracingContext"`
	RepositoryID      int        `json:"repositoryId"`
	Indexer           string     `json:"indexer"`
	Rank              *int       `json:"placeInQueue"`
}

// GetUploadByID returns an upload by its identifier and boolean flag indicating its existence.
func (db *dbImpl) GetUploadByID(ctx context.Context, id int) (Upload, bool, error) {
	query := `
		SELECT
			u.id,
			u.commit,
			u.root,
			u.visible_at_tip,
			u.uploaded_at,
			u.state,
			u.failure_summary,
			u.failure_stacktrace,
			u.started_at,
			u.finished_at,
			u.tracing_context,
			u.repository_id,
			u.indexer,
			s.rank
		FROM lsif_uploads u
		LEFT JOIN (
			SELECT r.id, RANK() OVER (ORDER BY r.uploaded_at) as rank
			FROM lsif_uploads r
			WHERE r.state = 'queued'
		) s
		ON u.id = s.id
		WHERE u.id = %s
	`

	upload, err := scanUpload(db.queryRow(ctx, sqlf.Sprintf(query, id)))
	if err != nil {
		return Upload{}, false, ignoreErrNoRows(err)
	}

	return upload, true, nil
}

// GetUploadsByRepo returns a list of uploads for a particular repo and the total count of records matching the given conditions.
func (db *dbImpl) GetUploadsByRepo(ctx context.Context, repositoryID int, state, term string, visibleAtTip bool, limit, offset int) (_ []Upload, _ int, err error) {
	tw, err := db.beginTx(ctx)
	if err != nil {
		return nil, 0, err
	}
	defer func() {
		err = closeTx(tw.tx, err)
	}()

	var conds []*sqlf.Query
	conds = append(conds, sqlf.Sprintf("u.repository_id = %s", repositoryID))
	if state != "" {
		conds = append(conds, sqlf.Sprintf("u.state = %s", state))
	}
	if term != "" {
		conds = append(conds, makeSearchCondition(term))
	}
	if visibleAtTip {
		conds = append(conds, sqlf.Sprintf("u.visible_at_tip = true"))
	}

	countQuery := `SELECT COUNT(1) FROM lsif_uploads u WHERE %s`
	count, err := scanInt(tw.queryRow(ctx, sqlf.Sprintf(countQuery, sqlf.Join(conds, " AND "))))
	if err != nil {
		return nil, 0, err
	}

	query := `
		SELECT
			u.id,
			u.commit,
			u.root,
			u.visible_at_tip,
			u.uploaded_at,
			u.state,
			u.failure_summary,
			u.failure_stacktrace,
			u.started_at,
			u.finished_at,
			u.tracing_context,
			u.repository_id,
			u.indexer,
			s.rank
		FROM lsif_uploads u
		LEFT JOIN (
			SELECT r.id, RANK() OVER (ORDER BY r.uploaded_at) as rank
			FROM lsif_uploads r
			WHERE r.state = 'queued'
		) s
		ON u.id = s.id
		WHERE %s ORDER BY uploaded_at DESC LIMIT %d OFFSET %d
	`

	uploads, err := scanUploads(tw.query(ctx, sqlf.Sprintf(query, sqlf.Join(conds, " AND "), limit, offset)))
	if err != nil {
		return nil, 0, err
	}

	return uploads, count, nil
}

// makeSearchCondition returns a disjunction of LIKE clauses against all searchable columns of an upload.
func makeSearchCondition(term string) *sqlf.Query {
	searchableColumns := []string{
		"commit",
		"root",
		"indexer",
		"failure_summary",
		"failure_stacktrace",
	}

	var termConds []*sqlf.Query
	for _, column := range searchableColumns {
		termConds = append(termConds, sqlf.Sprintf("u."+column+" LIKE %s", "%"+term+"%"))
	}

	return sqlf.Sprintf("(%s)", sqlf.Join(termConds, " OR "))
}

// Enqueue inserts a new upload with a "queued" state, returning its identifier and a TxCloser that must be closed to commit the transaction.
func (db *dbImpl) Enqueue(ctx context.Context, commit, root, tracingContext string, repositoryID int, indexerName string) (_ int, _ TxCloser, err error) {
	tw, err := db.beginTx(ctx)
	if err != nil {
		return 0, nil, err
	}
	defer func() {
		if err != nil {
			err = closeTx(tw.tx, err)
		}
	}()

	query := `
		INSERT INTO lsif_uploads (commit, root, tracing_context, repository_id, indexer)
		VALUES (%s, %s, %s, %s, %s)
		RETURNING id
	`

	id, err := scanInt(tw.queryRow(ctx, sqlf.Sprintf(query, commit, root, tracingContext, repositoryID, indexerName)))
	if err != nil {
		return 0, nil, err
	}

	return id, &txCloser{tw.tx}, nil
}

// GetStates returns the states for the uploads with the given identifiers.
func (db *dbImpl) GetStates(ctx context.Context, ids []int) (map[int]string, error) {
	query := `SELECT id, state FROM lsif_uploads WHERE id IN (%s)`
	return scanStates(db.query(ctx, sqlf.Sprintf(query, sqlf.Join(intsToQueries(ids), ", "))))
}

// DeleteUploadByID deletes an upload by its identifier. If the upload was visible at the tip of its repository's default branch,
// the visibility of all uploads for that repository are recalculated. The given function is expected to return the newest commit
// on the default branch when invoked.
func (db *dbImpl) DeleteUploadByID(ctx context.Context, id int, getTipCommit func(repositoryID int) (string, error)) (_ bool, err error) {
	tw, err := db.beginTx(ctx)
	if err != nil {
		return false, err
	}
	defer func() {
		err = closeTx(tw.tx, err)
	}()

	query := `
		DELETE FROM lsif_uploads
		WHERE id = %s
		RETURNING repository_id, visible_at_tip
	`

	repositoryID, visibleAtTip, err := scanVisibility(tw.queryRow(ctx, sqlf.Sprintf(query, id)))
	if err != nil {
		return false, ignoreErrNoRows(err)
	}

	if !visibleAtTip {
		return true, nil
	}

	tipCommit, err := getTipCommit(repositoryID)
	if err != nil {
		return false, err
	}

	if err := db.updateDumpsVisibleFromTip(ctx, tw, repositoryID, tipCommit); err != nil {
		return false, err
	}

	return true, nil
}

// ResetStalled moves all unlocked uploads processing for more than `StalledUploadMaxAge` back to the queued state.
// This method returns a list of updated upload identifiers.
func (db *dbImpl) ResetStalled(ctx context.Context, now time.Time) ([]int, error) {
	query := `
		UPDATE lsif_uploads u SET state = 'queued', started_at = null WHERE id = ANY(
			SELECT id FROM lsif_uploads
			WHERE state = 'processing' AND %s - started_at > (%s * interval '1 second')
			FOR UPDATE SKIP LOCKED
		)
		RETURNING u.id
	`

	ids, err := scanInts(db.query(ctx, sqlf.Sprintf(query, now.UTC(), StalledUploadMaxAge/time.Second)))
	if err != nil {
		return nil, err
	}

	return ids, nil
}
